anndata/_config.py

```diff
@@ -0,0 +1,240 @@
+            return doc
+        else:
+            return "\n".join(
+                [self.describe(k, print_description) for k in self._registered_options]
+            )
+
+    def deprecate(self, option: str, removal_version: str, message: str | None = None):
+        """Deprecate options with a message at a version.
+
+        Parameters
+        ----------
+        option
+            Which option should be deprecated.
+        removal_version
+            The version targeted for removal.
+        message
+            A custom message.
+        """
+        self._deprecated_options[option] = DeprecatedOption(
+            option, message, removal_version
+        )
+
+    def register(
+        self,
+        option: str,
+        default_value: object,
+        doc: str,
+        validator: Callable[[T], None],
+    ):
+        """Register an option so it can be set/described etc. by end-users
+
+        Parameters
+        ----------
+        option
+            Option to be set.
+        default_value
+            Default value with which to set the option.
+        doc
+            Docstring for the option
+        validator
+            A function which asserts that the option's value is valid.
+        """
+        validator(default_value)
+        self._registered_options[option] = RegisteredOption(
+            option, default_value, doc, validator
+        )
+        self._config[option] = default_value
+
+    def __setitem__(self, option: str, val: object):
+        """
+        Set an option to a value.  To see the allowed options to be set and their description,
+        use describe_option.
+
+        Parameters
+        ----------
+        option
+            Option to be set.
+        val
+            Value with which to set the option.
+
+        Raises
+        ------
+        KeyError
+            If the option has not been registered, this function will raise an error.
+        """
+        if option not in self._registered_options:
+            raise KeyError(
+                f"{option} is not an available option for anndata.\
+                Please open an issue if you believe this is a mistake."
+            )
+        registered_option = self._registered_options[option]
+        registered_option.validator(val)
+        self._config[option] = val
+
+    def __getitem__(self, option: str) -> object:
+        """
+        Gets the option's value.
+
+        Parameters
+        ----------
+        option
+            Option to be got.
+
+        Returns
+        -------
+        object
+            Value of the option.
+        """
+        return self._config[option]
+
+    def reset(self, option: str):
+        """
+        Resets an option to its default value.
+
+        Parameters
+        ----------
+        option
+            The option to be reset.
+        """
+        self._config[option] = self._registered_options[option].default_value
+
+    @contextmanager
+    def override(self, **overrides):
```

Comment:
This doesn’t support autocompletion, how can we fix that?

Reply:
Ok so I have spent quite a long time on this to no avail.  I have tried many different things.  I will need to look at this again on Monday.  If this is a breaking point, we may have to go the scanpy route (i.e., explicit attributes/keyword/positional arguments).  Something I realized while working through this, though, is that this solution _is generalizable_ in that we could migrate scanpy towards this.  But with the positional/keyword args way of doing things (i.e., what scanpy does), I don't think we can really support inheritance/re-usability.  But maybe I am missing something - I think there are ways of passing documentation up an inheritance tree, but I am not sure about types.

Reply:
you managed to do that now with `update_override_function`, no?

Reply:
This works for non-static analyzers, like IPython, but not vscode, which uses static analysis.